### PR TITLE
Benchmark Clad and Enzyme

### PR DESCRIPTION
**Pull request by Nirhar**
We compare the performance of Reverse Mode in Clad and Enzyme for 4 different kinds of functions:
1. Sum of elements of an array
2. Product of elements of an array
3. Sum of elements of an array multiplied with 2 scalars
4. The Gaussian density function

### CODE REVIEW

File: benchmark/EnzymeCladComparison.cpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
+#include "benchmark/benchmark.h"
+
+// FIXME: If we move this before benchmark.h we have tons of errors due to a bug
+#include "clad/Differentiator/Differentiator.h"
+
+#include "BenchmarkedFunctions.h"
+
+// Compare the execution of Reverse Mode Enzyme and Clad.
+
+// FIXME: Copy made because when we run clad::gradient on the same function
```
Comment (vgvassilev) [inline benchmark/EnzymeCladComparison.cpp:10]: Wasn't that addressed by #492

Comment (Nirhar) [reply benchmark/EnzymeCladComparison.cpp]: yes, have removed the copy functions now

File: benchmark/EnzymeCladComparison.cpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+  double dx[5];
+  double dp[5];
+  double ds;
+  int ddim;
+
+  for (auto _ : state) {
+    dfdp_grad.execute(x, p, /*sigma*/ 2, dim, dx, dp, &ds, &ddim);
+    for (int i = 0; i < dim; i++) {
+      // benchmark::DoNotOptimize(sum += result_ref[i]);
```
Comment (vgvassilev) [inline benchmark/EnzymeCladComparison.cpp:138]: Does uncommenting this lines change anything for the benchmarks?

Comment (Nirhar) [reply benchmark/EnzymeCladComparison.cpp:138]: No, the overall performance difference of the gauss function looks the same to me.

Comment (Nirhar) [reply benchmark/EnzymeCladComparison.cpp:138]: So should I remove these `benchmark::DoNotOptimize` statements from everywhere?

Comment (vgvassilev) [reply benchmark/EnzymeCladComparison.cpp:138]: Yes.

**Review by vgvassilev — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**